test/Validate.CFG.cpp

```diff
@@ -581,10 +580,133 @@ TEST_F(ValidateCFG, UnreachableBranch) {
   CompileSuccessfully(str);
   ASSERT_EQ(SPV_SUCCESS, ValidateInstructions());
 }
-/// TODO(umar): Empty function
-/// TODO(umar): Single block loops
-/// TODO(umar): Nested loops
-/// TODO(umar): Nested selection
+
+TEST_F(ValidateCFG, EmptyFunction) {
+  string str = header + types_consts +
+               "%func    = OpFunction %voidt None %funct\n" + "OpFunctionEnd\n";
+
+  CompileSuccessfully(str);
+  ASSERT_EQ(SPV_SUCCESS, ValidateInstructions());
+}
+
+TEST_F(ValidateCFG, SingleBlockLoop) {
+  Block entry("entry");
+  Block loop("loop", SpvOpBranchConditional);
+  Block exit("exit", SpvOpReturn);
+
+  loop.setBody(
+      "%cond    = OpSLessThan %intt %one %two\n"
+      "OpLoopMerge %exit %loop None\n");
+
+  string str =
+      header + types_consts + "%func    = OpFunction %voidt None %funct\n";
+
+  str += entry >> loop;
+  str += loop >> vector<Block>({loop, exit});
+  str += exit;
+  str += "OpFunctionEnd";
+
+  CompileSuccessfully(str);
+  ASSERT_EQ(SPV_SUCCESS, ValidateInstructions());
+}
+
+TEST_F(ValidateCFG, NestedLoops) {
+  Block entry("entry");
+  Block loop1("loop1");
+  Block loop2("loop2", SpvOpBranchConditional);
+  Block loop2_merge("loop2_merge");
+  Block loop1_merge("loop1_merge", SpvOpBranchConditional);
+  Block exit("exit", SpvOpReturn);
+
+  loop1.setBody(
+      "%cond    = OpSLessThan %intt %one %two\n"
+      "OpLoopMerge %loop1_merge %loop2 None\n");
+
+  loop2.setBody("OpLoopMerge %loop2_merge %loop2 None\n");
+
+  string str =
+      header + types_consts + "%func    = OpFunction %voidt None %funct\n";
+
+  str += entry >> loop1;
+  str += loop1 >> loop2;
+  str += loop2 >> vector<Block>({loop2, loop2_merge});
+  str += loop2_merge >> loop1_merge;
+  str += loop1_merge >> vector<Block>({loop1, exit});
```

Comment:
I'm afraid this is subtly wrong: `loop1_merge` branches to `exit`, which is outside the continue construct.  But `loop1_merge` isn't allowed to branch outside because of [this restriction](https://www.khronos.org/registry/spir-v/specs/1.1/SPIRV.html#_a_id_structuredcontrolflow_a_structured_control_flow) (look for  "the only blocks in a construct that can branch outside the construct are").

I think a quick way to fix this is to modify `loop1`'s continue target to be itself, i.e.:
```
loop1.setBody(
       "%cond    = OpSLessThan %intt %one %two\n"
       "OpLoopMerge %loop1_merge %loop1 None\n");
```